core/include/kernel/generic_pm.h

```diff
@@ -0,0 +1,78 @@
+/* SPDX-License-Identifier: BSD-2-Clause */
+/*
+ * Copyright (c) 2018, Linaro Limited
+ */
+
+#ifndef __GENERIC_PM_H
+#define __GENERIC_PM_H
+
+#include <stdbool.h>
+#include <stdint.h>
+
+#define PM_SUBSYS_DEFAULT		0
+
+enum pm_op {
+	PM_OP_SUSPEND = 0,
+	PM_OP_RESUME = 1,
+};
+
+enum pm_cb_level {
+	PM_CB_LEVEL_DRIVER,
+	PM_CB_LEVEL_CORE_SERVICE,
+	PM_CB_LEVELS
+};
+
+/*
+ * register_pm_cb2 - Register a callback for a subsystem suspend/resume
+ *
+ * @callback: Registered callback function
+ * @handle: Registered cookie used as callback argument
+ * @cb_level: Ordering for the callback functions call
+ * @pm_subsys: ID for the suspended subsystem for the callback
+ *
+ * Drivers and services can register a callback for the suspend and resume
+ * sequences and a private cookie passed as argument to the callback.
+ * The same callback is used for suspend and resume. First argument of the
+ * callback defines whether the device shall suspend or resume.
+ *
+ * Driver should tag its callback resource as unpaged (i.e KEEP_PAGER())
+ * since the callback is called from an unpaged execution context.
+ *
+ * Callback shall return 0 on sucess and return 1 in case of failure. This
+ * may be used during a failing suspend sequence to call subsystem resume
+ * sequence.
+ *
+ * @pm_subsys defines the identifier which is used to notify generic PM
+ * of a subsystem suspend/resume. Callback registered for that subsystem
+ * will be suspended/resumed. Subsystem ID must be lower than UINT16_MAX.
+ */
+void register_pm_cb2(int (*callback)(enum pm_op op, void *handle),
+		     void *handle, enum pm_cb_level cb_level,
+		     unsigned int subsys);
+
+/*
+ * register_pm_cb - Register a driver callback for generic suspend/resume
+ *
+ * @callback: Registered callback function
+ * @handle: Registered cookie used as callback argument
+ */
+static inline void register_pm_cb(int (*callback)(enum pm_op op, void *handle),
+				  void *handle)
+{
+	register_pm_cb2(callback, handle,
+			PM_CB_LEVEL_DRIVER, PM_SUBSYS_DEFAULT);
+}
+
+/*
+ * pm_suspend_resume - Call registered suspend/resume callbacks
+ *
+ * In case of failure when suspending, platform can resume the target
+ * subsystem, only the still suspended components will be resumed.
+ *
+ * @pm_subsys: Target subsystem suspended or resumed or 0 if none
+ * @op: Either PM_OP_SUSPEND or PM_OP_RESUME
+ * Return 0 on success, 1 on failure of a callback.
+ */
+int pm_suspend_resume(unsigned int subsys, enum pm_op op);
```

Comment:
`pm_change_state()`
Also I wonder if we shouldn't return a `TEE_Result` instead.